geth/common/types.go

```diff
@@ -168,7 +169,34 @@ type SendTxArgs struct {
 	GasPrice *hexutil.Big    `json:"gasPrice"`
 	Value    *hexutil.Big    `json:"value"`
 	Nonce    *hexutil.Uint64 `json:"nonce"`
-	Input    hexutil.Bytes   `json:"input"`
+	// We keep both "input" and "data" for backward compatibility.
+	// "input" is a preferred field.
+	// see `vendor/github.com/ethereum/go-ethereum/internal/ethapi/api.go:1107`
+	Input hexutil.Bytes `json:"input"`
+	Data  hexutil.Bytes `json:"data"`
+}
+
+func (args SendTxArgs) Valid() bool {
+	// if at least one of the fields is empty, it is a valid struct
+	if isNilOrEmpty(args.Input) || isNilOrEmpty(args.Data) {
+		return true
+	}
+
+	// we only allow both fields to present if they have the same data
+	return bytes.Equal(args.Input, args.Data)
+}
+
+// GetInput returns either Input or Data field's value dependent on what is filled.
+func (args SendTxArgs) GetInput() hexutil.Bytes {
+	if !isNilOrEmpty(args.Input) {
+		return args.Input
+	}
+
+	return args.Data
+}
+
+func isNilOrEmpty(bytes hexutil.Bytes) bool {
+	return bytes == nil && len(bytes) == 0
```

Comment:
So it should be `bytes == nil || len(bytes) == 0`, right?

Reply:
good catch 👍